bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/communication/MonopriceAudioConnector.java

```diff
@@ -0,0 +1,260 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.monopriceaudio.internal.MonopriceAudioException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class for communicating with the MonopriceAudio device
+ *
+ * @author Laurent Garnier - Initial contribution
+ * @author Michael Lobstein - Adapted for the MonopriceAudio binding
+ */
+@NonNullByDefault
+public abstract class MonopriceAudioConnector {
+    public static final String READ_ERROR = "Command Error.";
+
+    // Message types
+    public static final String KEY_ZONE_UPDATE = "zone_update";
+    // Special keys used by the binding
+    public static final String KEY_ERROR = "error";
+    public static final String MSG_VALUE_ON = "on";
+
+    private static final Pattern PATTERN = Pattern.compile("^.*#>(\\d{22})$", Pattern.DOTALL);
+
+    private final Logger logger = LoggerFactory.getLogger(MonopriceAudioConnector.class);
+
+    /** The output stream */
+    protected @Nullable OutputStream dataOut;
+
+    /** The input stream */
+    protected @Nullable InputStream dataIn;
+
+    /** true if the connection is established, false if not */
+    private boolean connected;
+
+    private @Nullable Thread readerThread;
+
+    private List<MonopriceAudioMessageEventListener> listeners = new ArrayList<>();
+
+    /**
+     * Get whether the connection is established or not
+     *
+     * @return true if the connection is established
+     */
+    public boolean isConnected() {
+        return connected;
+    }
+
+    /**
+     * Set whether the connection is established or not
+     *
+     * @param connected true if the connection is established
+     */
+    protected void setConnected(boolean connected) {
+        this.connected = connected;
+    }
+
+    /**
+     * Set the thread that handles the feedback messages
+     *
+     * @param readerThread the thread
+     */
+    protected void setReaderThread(Thread readerThread) {
+        this.readerThread = readerThread;
+    }
+
+    /**
+     * Open the connection with the MonopriceAudio device
+     *
+     * @throws MonopriceAudioException - In case of any problem
+     */
+    public abstract void open() throws MonopriceAudioException;
+
+    /**
+     * Close the connection with the MonopriceAudio device
+     */
+    public abstract void close();
+
+    /**
+     * Stop the thread that handles the feedback messages and close the opened input and output streams
+     */
+    protected void cleanup() {
+        Thread readerThread = this.readerThread;
+        if (readerThread != null) {
+            readerThread.interrupt();
+            try {
+                readerThread.join();
```

Comment:
It is dangerous to wait indefinitely for the join, so add a reasonable timeout and log a warning if that timeout is exceeded.